modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/RebalanceStatisticsUtils.java

```diff
@@ -79,536 +67,267 @@ private RebalanceStatisticsUtils() {
+     * IgniteSystemProperties#IGNITE_WRITE_REBALANCE_PARTITION_DISTRIBUTION_THRESHOLD}.
      *
-     * @return Is print partitions distribution enabled.
+     * @param d Duration of rebalance, in milliseconds.
+     * @return {@code True} if printing partitions distribution is available.
      */
-    public static boolean printPartitionsDistribution() {
-        return getBoolean(IGNITE_WRITE_REBALANCE_PARTITION_STATISTICS, false);
+    public static boolean availablePrintPartitionsDistribution(long d) {
+        return d >= getLong(IGNITE_WRITE_REBALANCE_PARTITION_DISTRIBUTION_THRESHOLD, MINUTES.toMillis(10));
     }
 
     /**
-     * Return rebalance statistics. Required to call this method if
-     * {@link #printRebalanceStatistics()} == true.
-     * <p/>
-     * Flag {@code finish} should reflect was full rebalance finished or not.
-     * <br/>
-     * If {@code finish} == true then expected {@code rebFutrs} contains
-     * successful or not {@code RebalanceFuture} per cache group, else expected
-     * {@code rebFutrs} contains only one successful {@code RebalanceFuture}.
-     * <br/>
-     * If {@code finish} == true then print total statistics.
-     * <p/>
-     * Partition distribution is printed only for last success rebalance,
-     * per cache group.
+     * Creating a string representation of group cache rebalance statistics for
+     * logging.
      *
-     * @param finish Is the whole rebalance finished or not.
-     * @param rebFutrs Involved in rebalance, require not null.
-     * @return String with printed rebalance statistics.
-     * @throws IgniteCheckedException Could be thrown while getting result of
-     *      {@code RebalanceFuture}.
-     * @see RebalanceFuture RebalanceFuture
+     * @param cacheGrpCtx Cache group context.
+     * @param stat Rebalance statistics.
+     * @param suc Flag for successful rebalancing.
+     * @param top Topology version.
+     * @return String representation of rebalance statistics for cache group.
      */
-    public static String rebalanceStatistics(
-        final boolean finish,
-        final Map<CacheGroupContext, Collection<RebalanceFuture>> rebFutrs
-    ) throws IgniteCheckedException {
-        assert nonNull(rebFutrs);
-        assert printRebalanceStatistics() : "Can't print statistics";
-
-        AtomicInteger nodeCnt = new AtomicInteger();
+    public static String cacheGroupRebalanceStatistics(
+        CacheGroupContext cacheGrpCtx,
+        CacheGroupRebalanceStatistics stat,
+        boolean suc,
+        AffinityTopologyVersion top
+    ) {
+        SB sb = new SB();
 
-        Map<ClusterNode, Integer> nodeAliases = toRebalanceFutureStream(rebFutrs)
-            .flatMap(future -> future.stat.msgStats.keySet().stream())
-            .distinct()
-            .collect(toMap(identity(), node -> nodeCnt.getAndIncrement()));
+        sb.a("Information per cache group (").a(suc ? "successful" : "interrupted").a(" rebalance): [")
```

Comment:
Let's start this log message with word "rebalance".
It should be clear what's it about from the first glance.